components/brave_wallet/browser/solana_message.cc

```diff
@@ -13,30 +13,174 @@
 #include "base/ranges/algorithm.h"
 #include "base/strings/string_number_conversions.h"
 #include "base/values.h"
+#include "brave/components/brave_wallet/browser/solana_compiled_instruction.h"
 #include "brave/components/brave_wallet/common/brave_wallet_constants.h"
 #include "brave/components/brave_wallet/common/solana_utils.h"
 
 namespace brave_wallet {
 
-SolanaMessage::SolanaMessage(const std::string& recent_blockhash,
-                             uint64_t last_valid_block_height,
-                             const std::string& fee_payer,
-                             std::vector<SolanaInstruction>&& instructions)
-    : recent_blockhash_(recent_blockhash),
+namespace {
+
+const uint8_t kV0MessagePrefix = 0x80;
+const uint8_t kVersionPrefixMask = 0x7f;
+
+bool MaybeAddVersionPrefix(mojom::SolanaMessageVersion version,
+                           std::vector<uint8_t>* message_bytes) {
+  DCHECK(message_bytes);
+
+  if (version == mojom::SolanaMessageVersion::kV0) {
+    message_bytes->emplace_back(kV0MessagePrefix);
+  } else if (version == mojom::SolanaMessageVersion::kLegacy) {
+    // Do nothing.
+  } else {  // Version not supported yet.
+    return false;
+  }
+
+  return true;
+}
+
+// Deserialize message header from a serialized message.
+// Returns the bytes_index after message header part, deserialized version and
+// message header.
+absl::optional<
+    std::tuple<size_t, mojom::SolanaMessageVersion, SolanaMessageHeader>>
+DeserializeMessageHeader(const std::vector<uint8_t>& bytes) {
+  // Check version
+  if (bytes.size() < 1) {
```

Comment:
Every transaction should have a message header section AFAIK.

If bytes.size is 0 (I assume this is what you meant about bytes is actually null?), `SolanaMessage::Deserialize` would fail and the API call would error out, it's pretty early as this is the first thing to deserialize.
I might misunderstand what you meant about bytes is actually null tho.

Reply:
Ahh ignore this one, I made a mistake. I misread the [second paragraph here](https://docs.solana.com/proposals/versioned-transactions#versioned-transactions) to assume that the upper bit modification was an upper byte and there were times when that wouldn't exist. This looks fine to me.